### PR TITLE
add the option for fast downward grounding

### LINKED ISSUE #885: implement smarter operator grounding techniques
mainly we should be able to leverage static preconditions to avoid grounding a bunch of never-applicable operators.

the forthcoming doors environment will be a testbed for this.

### PR DESCRIPTION
**Pull request by ronuchit**
Some results:

_with new grounding_
`python src/main.py --env blocks --approach oracle --seed 0 --sesame_grounder fd_translator`
```
Tasks solved: 49 / 50
Average time for successes: 0.12967 seconds
Test results: defaultdict(<class 'float'>, {'num_solved': 49, 'num_total': 50, 'avg_suc_time': 0.12966873694439324, 'min_skeletons_optimized': 1.0, 'max_skeletons_optimized': 4.0, 'num_execution_failures': 0, 'num_solve_failures': 1, 'avg_num_skeletons_optimized': 1.0612244897959184, 'avg_num_nodes_expanded': 333.9591836734694, 'avg_num_nodes_created': 939.7551020408164, 'avg_num_nsrts': 4.0, 'avg_num_preds': 5.0, 'avg_plan_length': 7.959183673469388, 'avg_num_failures_discovered': 0.0, 'num_offline_transitions': 0, 'num_online_transitions': 0, 'query_cost': 0.0, 'learning_time': 0.0})
```

`python src/main.py --env doors --approach oracle --seed 0 --sesame_grounder fd_translator`
```
Tasks solved: 50 / 50
Average time for successes: 0.79166 seconds
Test results: defaultdict(<class 'float'>, {'num_solved': 50, 'num_total': 50, 'avg_suc_time': 0.7916550159454345, 'min_skeletons_optimized': 1.0, 'max_skeletons_optimized': 1.0, 'num_execution_failures': 0, 'num_solve_failures': 0, 'avg_num_skeletons_optimized': 1.0, 'avg_num_nodes_expanded': 13.48, 'avg_num_nodes_created': 70.96, 'avg_num_nsrts': 4.0, 'avg_num_preds': 7.0, 'avg_plan_length': 7.26, 'avg_num_failures_discovered': 0.0, 'num_offline_transitions': 0, 'num_online_transitions': 0, 'query_cost': 0.0, 'learning_time': 0.0})
```

_with old grounding_
`python src/main.py --env blocks --approach oracle --seed 0 --sesame_grounder naive`
```
Tasks solved: 49 / 50
Average time for successes: 0.11200 seconds
Test results: defaultdict(<class 'float'>, {'num_solved': 49, 'num_total': 50, 'avg_suc_time': 0.1120033750728685, 'min_skeletons_optimized': 1.0, 'max_skeletons_optimized': 4.0, 'num_execution_failures': 0, 'num_solve_failures': 1, 'avg_num_skeletons_optimized': 1.0612244897959184, 'avg_num_nodes_expanded': 333.9591836734694, 'avg_num_nodes_created': 939.7551020408164, 'avg_num_nsrts': 4.0, 'avg_num_preds': 5.0, 'avg_plan_length': 7.959183673469388, 'avg_num_failures_discovered': 0.0, 'num_offline_transitions': 0, 'num_online_transitions': 0, 'query_cost': 0.0, 'learning_time': 0.0})
```

`python src/main.py --env doors --approach oracle --seed 0 --sesame_grounder naive`
```
Tasks solved: 50 / 50
Average time for successes: 1.24272 seconds
Test results: defaultdict(<class 'float'>, {'num_solved': 50, 'num_total': 50, 'avg_suc_time': 1.2427226734161376, 'min_skeletons_optimized': 1.0, 'max_skeletons_optimized': 1.0, 'num_execution_failures': 0, 'num_solve_failures': 0, 'avg_num_skeletons_optimized': 1.0, 'avg_num_nodes_expanded': 14.78, 'avg_num_nodes_created': 84.82, 'avg_num_nsrts': 4.0, 'avg_num_preds': 7.0, 'avg_plan_length': 7.26, 'avg_num_failures_discovered': 0.0, 'num_offline_transitions': 0, 'num_online_transitions': 0, 'query_cost': 0.0, 'learning_time': 0.0})
```

TL;DR: `fd_translator` shows no difference in blocks, ~30% faster in doors

closes #885

### CODE REVIEW

File: src/planning.py [hunk 1]
```diff
@@ -62,16 +64,34 @@ def sesame_plan(
...
-            if time.time() - start_time > timeout:
-                raise PlanningTimeout("Planning timed out in grounding!")
+    if CFG.sesame_grounder == "naive":
+        ground_nsrts = []
+        for nsrt in sorted(nsrts):
+            for ground_nsrt in utils.all_ground_nsrts(nsrt, objects):
+                ground_nsrts.append(ground_nsrt)
+                if time.time() - start_time > timeout:
+                    raise PlanningTimeout("Planning timed out in grounding!")
+    elif CFG.sesame_grounder == "fd_translator":
```
Comment (tomsilver) [inline src/planning.py:77]: can you add some comments here, or in settings.py, about when people should use one vs the other? probably mentioning the noop issue is worthwhile

File: src/planning.py [hunk 2]
```diff
@@ -62,16 +64,34 @@ def sesame_plan(
...
-        for ground_nsrt in utils.all_ground_nsrts(nsrt, objects):
-            ground_nsrts.append(ground_nsrt)
-            if time.time() - start_time > timeout:
-                raise PlanningTimeout("Planning timed out in grounding!")
+    if CFG.sesame_grounder == "naive":
+        ground_nsrts = []
+        for nsrt in sorted(nsrts):
+            for ground_nsrt in utils.all_ground_nsrts(nsrt, objects):
+                ground_nsrts.append(ground_nsrt)
+                if time.time() - start_time > timeout:
```
Comment (tomsilver) [inline src/planning.py:75]: shouldn't this timeout apply to either grounding method?

Comment (tomsilver) [reply src/planning.py:75]: I guess there's no easy way to interrupt grounding within FD (and btw this is another thing worth warning about in a comment), but, we should check once whether we've timed out after grounding

File: src/planning.py [hunk 3]
```diff
@@ -62,16 +64,34 @@ def sesame_plan(
...
+                    raise PlanningTimeout("Planning timed out in grounding!")
+    elif CFG.sesame_grounder == "fd_translator":
+        nsrt_name_to_nsrt = {nsrt.name.lower(): nsrt for nsrt in nsrts}
```
Comment (tomsilver) [inline src/planning.py:78]: make this all a utility function in utils.py? mostly for cleanliness, but maybe one day for reuse

**Review by tomsilver — APPROVED** [review]
a few comments but looks good. can you also add some timing results to the PR description?

File: tests/test_planning.py [hunk 4]
```diff
@@ -18,33 +20,43 @@
...
 
 
-@pytest.mark.parametrize("sesame_check_expected_atoms", [True, False])
-def test_sesame_plan(sesame_check_expected_atoms):
+@pytest.mark.parametrize(
+    "sesame_check_expected_atoms,sesame_grounder,expectation",
+    [(True, "naive", does_not_raise()), (False, "naive", does_not_raise()),
+     (True, "fd_translator", does_not_raise()),
+     (True, "not a real grounder", pytest.raises(ValueError))])
+def test_sesame_plan(sesame_check_expected_atoms, sesame_grounder,
```
Comment (ronuchit) [inline tests/test_planning.py:28]: @tomsilver this test provides a neat example of a pattern that we can use elsewhere going forward -- parametrization with some cases raising an exception
